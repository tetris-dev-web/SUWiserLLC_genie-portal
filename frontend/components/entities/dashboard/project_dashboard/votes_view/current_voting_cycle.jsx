import React from 'react';
import * as d3 from 'd3';
import { connect } from 'react-redux';
import VoteShiftTool from './vote_shift_tool';

class VotesView2 extends React.Component{
  constructor(props){
    super(props);
    this.state = {
      showVoteShiftTool: false,
      selectedProject: null
    };

    this.handleClick = this.handleClick.bind(this);
  }

  componentDidUpdate() {
    // needs a lifecycle method when props is changed
  }

  componentDidMount(){
    const { maxValuation, capitalRaised } = this.props;
    this.svg = d3.select(".votes-view")
      .append("svg")
      .classed("votes-view-svg", true)
      .attr("width", "100%")
      .attr("height", maxValuation/this.props.scalingConstant);

    this.svg.append("g")
      .append("rect")
      .classed("current-cycle-capital", true)
      .attr("width", "100%")
      .attr("height", capitalRaised / this.props.scalingConstant)
      .attr("fill", "#aa7a60")
      .style("opacity", .5);

      this.svg.append('text')
      .text(() => {
        return this.formatNumber(capitalRaised);
      })
      .style('fill', '#fff')
      .style('font-size', '18px')
      .style('text-anchor', 'start')
      .attr('x', '4%')
      .attr('y', '30%');

      this.svg.append('text')
      .text("capital being")
      .style('fill', '#fff')
      .style('font-size', '18px')
      // .style('text-anchor', 'end')
      .attr('x', '90%')
      .attr('y', '40%')
      .append('tspan')
      .text('raised')
      .attr('dy', '1.2em')
      // .attr('text-anchor', 'middle')
      .attr('dx', '-3em')
      .style('fill', '#fff')
      .style('font-size', '18px')
      // .attr('x', '90%')
      // .attr('y', '40%')
      .text('raised');


    this.setUp();
  }

  formatNumber(number){
    return `$ ${Number(number/1000.0).toLocaleString()} k`;
  }

  setUp(){
    const { capitalRaised } = this.props;
    const projects = this.processProjectData();

    const valuationRect = this.svg.selectAll(".valuation-rect").remove();
    valuationRect.data(projects)
      .enter()
      .append("rect")
      .attr("class", "valuation-rect")
      .attr("width", project => `${project.projectWidth}%`)
      .attr("height", project => project.projectValutionHeight)
      .attr("fill", project => project.fill)
      .attr("x", project => `${project.projectStartX}%`)
      .attr("y", project => project.projectValutionStartY)
      .attr("opacity", .3)
      // .style("stroke", "white")
      // .style("stroke-width", 2)
      .on("mouseover", this.handleMouseOver())
      .on("mouseout", this.handleMouseOut())
      .on("click", this.handleClick);

    const capRequiredRect = this.svg.selectAll(".capital-required-rect").remove();
    capRequiredRect.data(projects)
      .enter()
      .append("rect")
      .attr("class", "capital-required-rect")
      .attr("width", project => `${project.projectWidth}%`)
      .attr("height", project => project.projectCapitalRequiredHeight)
      .attr("fill", project => project.fill)
      .attr("x", project => `${project.projectStartX}%`)
      .attr("y", project => project.projectCapitalRequiredStartY)
      .attr("opacity", 1)
      // .style("stroke", "white")
      // .style("stroke-width", 2)
      .on("mouseover", this.handleMouseOver())
      .on("mouseout", this.handleMouseOut())
      .on("click", this.handleClick);

    const votePercentageRect = this.svg.selectAll(".vote-percentage-rect").remove();
    votePercentageRect.data(projects)
      .enter()
      .append("rect")
      .attr("class", "vote-percentage-rect")
      .attr("width", project => `${project.projectWidth + 1}%`)
      .attr("height", 30)
      .attr("fill", project => project.fill)
      .attr("x", project => `${project.projectStartX - .5}%`)
      .attr("y", capitalRaised / this.props.scalingConstant);
      // .style("stroke", "white")
      // .style("stroke-width", 2);

    const votePercentageText = this.svg.selectAll(".vote-percentage-text").remove();
    votePercentageText.data(projects)
      .enter()
      .append("text")
      .attr("class", "vote-percentage-text")
      .attr("x", project => `${project.projectRectCenter}%`)
      .attr("y", capitalRaised / this.props.scalingConstant + 20)
      .style("font-size", "18px")
      .style("fill", "#fff")
      .style("text-anchor", "middle")
      .text(project => `${project.voteShare * 100}%`);
  }

  processProjectData(){
    const { capitalRaised, pitchedProjects } = this.props;
    const numberOfProjects = this.props.pitchedProjects.length;
    const percentOfScreen = 60;
    const projectWidthPercentage = percentOfScreen - (numberOfProjects - 1);
    let projectStartX = 20;

    return pitchedProjects.map(project => {
      const projectWidth = project.voteShare * projectWidthPercentage;
      const newProject = Object.assign({}, project, {
        projectStartX,
        projectWidth,
        fill: capitalRaised < project.capitalRequired ? "#aa7a60" : "#61aba9",
        projectValutionHeight: (project.valuation - project.capitalRequired) / this.props.scalingConstant,
        projectValutionStartY: (capitalRaised - project.valuation) / this.props.scalingConstant,
        projectCapitalRequiredHeight: project.capitalRequired / this.props.scalingConstant,
        projectCapitalRequiredStartY: (capitalRaised - project.capitalRequired) / this.props.scalingConstant,
        projectRectCenter: projectStartX + projectWidth / 2
      });
      projectStartX += (1 + projectWidth);
      return newProject;
    });
  }

  handleMouseOver(){
    const { maxValuation, capitalRaised } = this.props;
    const appendTextToRect = (textContent, textFill, textSize, textX, textY) => {
      this.svg.append("text")
        .attr("class", "rect-text")
        .attr("x", textX)
        .attr("y", textY)
        .attr("pointer-events", "none")
        .style("font-size", textSize)
        .style("fill", textFill)
        .style("text-anchor", "middle")
        .text(textContent);
    };

    const appendOutLines = (outlineX, outlineY, outlineWidth, outlineHeight) => {
      this.svg.append("rect")
        .attr("class", "rect-outline")
        .attr("x", outlineX)
        .attr("y", outlineY)
        .attr("width", outlineWidth)
        .attr("height", outlineHeight)
        .style("fill", "#fff");
    };

    return project => {
      appendTextToRect(project.valuation, project.fill, "12px", `${project.projectRectCenter}%`, project.projectValutionStartY + 15);
      appendTextToRect(project.capitalRequired, "white", "12px", `${project.projectRectCenter}%`, project.projectCapitalRequiredStartY + 15);
      appendTextToRect("valution", project.fill, "12px", `${project.projectRectCenter}%`, project.projectValutionStartY - 7);
      appendTextToRect("capital needs", "white", "12px", `${project.projectRectCenter}%`, project.projectCapitalRequiredStartY - 7);
      appendTextToRect(project.title, "#aa7a60", "15px", `${project.projectRectCenter}%`, -(maxValuation - capitalRaised) / this.props.scalingConstant * 1.5);
      appendOutLines(`${project.projectStartX}%`, project.projectValutionStartY, `${project.projectWidth}%`, 2);
      appendOutLines(`${project.projectStartX + project.projectWidth}%`, project.projectValutionStartY, 2, project.valuation / this.props.scalingConstant);
      appendOutLines(`${project.projectStartX + project.projectWidth}%`, capitalRaised / this.props.scalingConstant, "0.5%", 2);
      appendOutLines(`${project.projectStartX + project.projectWidth + 0.5}%`, capitalRaised / this.props.scalingConstant, 2, 30);
      appendOutLines(`${project.projectStartX - 0.5}%`, capitalRaised / this.props.scalingConstant + 28, `${project.projectWidth + 1}%`, 2);
      appendOutLines(`${project.projectStartX - 0.5}%`, capitalRaised / this.props.scalingConstant, 2, 30);
      appendOutLines(`${project.projectStartX - 0.5}%`, capitalRaised / this.props.scalingConstant, "0.5%", 2);
      appendOutLines(`${project.projectStartX}%`, project.projectValutionStartY + 2, 2, project.valuation / this.props.scalingConstant);
    };
  }

  handleMouseOut(){
    return () => {
      d3.selectAll(".rect-text").remove();
      d3.selectAll(".rect-outline").remove();
    };
  }

  handleClick(project) {
    const { showVoteShiftTool } = this.state;
    if (showVoteShiftTool) {
      this.setState({
        showVoteShiftTool: !showVoteShiftTool,
        selectedProject: null
      });
    } else {
      this.setState({
        showVoteShiftTool: !showVoteShiftTool,
        selectedProject: project
      });
    }
  }

  render(){
    const { maxValuation, capitalRaised } = this.props;
    return(
      <div className="votes-view" style={{marginTop: (maxValuation - capitalRaised) / this.props.scalingConstant * 2}}>
        {this.state.showVoteShiftTool && <VoteShiftTool project={this.state.selectedProject}/>}
      </div>
    );
  }
}


//CONTAINER

//is this data being used?
//
// const mapStateToProps = state => {
//     return({
//       pitchedProjects: [
//         {
//           valuation: 5000000,
//           capitalRequired: 3500000,
//           voteShare: .15,
//           title: "Ryan and Liam"
//         },
//         {
//           valuation: 2000000,
//           capitalRequired: 1500000,
//           voteShare: .15,
//           title: "Liam and Ryan"
//         },
//         {
//           valuation: 4000000,
//           capitalRequired: 2500000,
//           voteShare: .20,
//           title: "HamInn"
//         },
//         {
//           valuation: 5000000,
//           capitalRequired: 4000000,
//           voteShare: .30,
//           title: "Genesis"
//         },
//         {
//           valuation: 3500000,
//           capitalRequired: 2000000,
//           voteShare: .20,
//           title: "Penn Generator"
//         },
//       ].sort((a,b) => b.voteShare - a.voteShare),
//       maxValuation: 5000000,
//       capitalRaised: 3000000
//     });
// };

export default VotesView2;
