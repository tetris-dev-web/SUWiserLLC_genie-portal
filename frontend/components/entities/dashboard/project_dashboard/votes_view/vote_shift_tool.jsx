import React from 'react';
import './vote_shift_tool.scss';

// TODO make into object
const VOTE_BAR_WIDTH = 140;
const VOTE_BAR_HEIGHT = 25;
const VOTE_BAR_INNER_MARGIN = 5;
const VOTE_BAR_RADIUS = 8;
const VOTE_SHIFT_LINE_WIDTH = 5;
const VOTE_SHIFT_LINE_HEIGHT = 80;
const INNER_BAR_HEIGHT = VOTE_BAR_HEIGHT - 2 * VOTE_BAR_INNER_MARGIN;

class VoteShiftTool extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      showLogButton: false,
      blockchainLoading: false,
      newVotesPerProject: 0,
      newVotesNotDedicated: 0,
      voteBarAppliedWidth: 0,
      voteBarFreedUpWidth: 0,
      voteShiftLineLeft: 0
    };

    this.fetchVoteData = this.fetchVoteData.bind(this);
    this.handleDragStart = this.handleDragStart.bind(this);
    this.handleDragEnd = this.handleDragEnd.bind(this);
    this.handleDragging = this.handleDragging.bind(this);
    this.handleLogClick = this.handleLogClick.bind(this);
    this.populateState = this.populateState.bind(this);
    this.watchVoteChange = this.watchVoteChange.bind(this);
  }

  componentDidMount () {
    this.fetchVoteData();
    this.watchVoteChange();
    this.populateState();
  }

  componentDidUpdate (prevProps, prevState) {
    const { votesNotDedicated, votesPerProject } = this.props;
    if (
      votesNotDedicated !== prevProps.votesNotDedicated ||
      votesPerProject !== prevProps.votesPerProject
    ) {
      this.populateState();
    }
  }

  fetchVoteData () {
    const {
      fetchFreeVotes,
      fetchProjectVotes,
      fetchDemoInvestorFreeVotes,
      fetchDemoInvestorProjectVotes,
      projectContract,
      votingToken,
      account,
      selectedProject,
      // loggedInToMetaMask
    } = this.props;

    // if (loggedInToMetaMask) {
      // fetchFreeVotes(account, votingToken).then(() => {
      //   fetchProjectVotes(account, projectContract, selectedProject);
      // })
    // } else {
      fetchDemoInvestorFreeVotes().then(() => fetchDemoInvestorProjectVotes(selectedProject))
    // }
  }

  handleDragStart(e) {
    e.preventDefault();
    this.offsetX = e.clientX - this.shiftLine.getBoundingClientRect().left;

    document.addEventListener('mousemove', this.handleDragging);
    document.addEventListener('mouseup', this.handleDragEnd);
  }

  handleDragging(e) {
    const mouseX = e.clientX - this.voteBarContainer.getBoundingClientRect().left;
    let voteShiftLineLeft = mouseX - this.offsetX;

    if (voteShiftLineLeft < 2 * VOTE_BAR_INNER_MARGIN) {
      voteShiftLineLeft = 2 * VOTE_BAR_INNER_MARGIN;
    }

    if (voteShiftLineLeft > VOTE_BAR_WIDTH - 2 * VOTE_BAR_INNER_MARGIN - VOTE_SHIFT_LINE_WIDTH) {
      voteShiftLineLeft = VOTE_BAR_WIDTH - 2 * VOTE_BAR_INNER_MARGIN - VOTE_SHIFT_LINE_WIDTH;
    }

    const voteBarAppliedWidth = voteShiftLineLeft - 2 * VOTE_BAR_INNER_MARGIN;
    const voteBarFreedUpWidth = VOTE_BAR_WIDTH - voteShiftLineLeft - 2 * VOTE_BAR_INNER_MARGIN - VOTE_SHIFT_LINE_WIDTH;

    this.setState({
      newVotesPerProject: Math.ceil(voteBarAppliedWidth * this.votesPerPixel),
      newVotesNotDedicated: Math.floor(voteBarFreedUpWidth * this.votesPerPixel),
      voteBarAppliedWidth,
      voteBarFreedUpWidth,
      voteShiftLineLeft
    });
  }

  handleDragEnd() {
    const { votesPerProject, votesNotDedicated } = this.props;

    document.removeEventListener('mousemove', this.handleDragging);
    document.removeEventListener('mouseup', this.handleDragEnd);
    this.setState({ showLogButton: !(this.state.newVotesPerProject === votesPerProject && this.state.newVotesNotDedicated === votesNotDedicated) });
  }

  handleLogClick() {
    this.setState({ blockchainLoading: !this.state.blockchainLoading });

    const { account, votingInstance, selectedProject, votesPerProject, voteAndUpdateProjects, demoInvestorVoteAndUpdateProjects, projects, activation, projectLeaderTracker, web3, updateTransactionModal } = this.props;
    const { newVotesPerProject } = this.state;

    let type;
    let votes;
    if (newVotesPerProject > votesPerProject) {
      type = 'addVotes';
      votes = newVotesPerProject - votesPerProject;
    } else {
      type = 'removeVotes';
      votes = votesPerProject - newVotesPerProject;
    }
    // voteAndUpdateProjects(
    //   account,
    //   votes,
    //   type,
    //   votingInstance,
    //   selectedProject,
    //   projects,
    //   projectLeaderTracker,
    //   activation,
    //   web3
    // );

    demoInvestorVoteAndUpdateProjects(
      votes,
      type,
      selectedProject
    )

    updateTransactionModal({
      isOpen: true,
      title: "YOUR TRANSACTION HAS BEEN SENT",
      message: "It may take a few minutes for your transaction to be processed by the blockchain."
    })
  }

  handleVoteClick(vote) {
    const { votesPerProject, votesNotDedicated } = this.props;

    return () => {
      if (vote > 0 && this.totalVotes === this.state.newVotesPerProject ||
        vote < 0 && this.state.newVotesPerProject === 0) return;
      const voteShiftLineLeft = this.state.voteShiftLineLeft + vote / this.votesPerPixel;
      const voteBarAppliedWidth = voteShiftLineLeft - 2 * VOTE_BAR_INNER_MARGIN;
      const voteBarFreedUpWidth = VOTE_BAR_WIDTH - voteShiftLineLeft - 2 * VOTE_BAR_INNER_MARGIN - VOTE_SHIFT_LINE_WIDTH;

      this.setState({
        newVotesPerProject: this.state.newVotesPerProject + vote,
        newVotesNotDedicated: this.state.newVotesNotDedicated - vote,
        voteBarAppliedWidth,
        voteBarFreedUpWidth,
        voteShiftLineLeft,
        showLogButton: !(this.state.newVotesPerProject + vote === votesPerProject && this.state.newVotesNotDedicated - vote === votesNotDedicated)
      });
    };
  }

  populateState () {
    const { votesNotDedicated, votesPerProject } = this.props;
    // const votesPerProject = this.props.accountVotes[this.props.selectedProject];
    // const votesNotDedicated = this.props.accountVotes.freeVotes;
    this.offsetX = 0;
    this.totalVotes = votesNotDedicated + votesPerProject;
    this.votesPerPixel = this.totalVotes / (VOTE_BAR_WIDTH - 4 * VOTE_BAR_INNER_MARGIN - VOTE_SHIFT_LINE_WIDTH);

    const voteBarAppliedWidth = votesPerProject === 0 ? 0 : votesPerProject / this.votesPerPixel;
    const voteBarFreedUpWidth = votesNotDedicated === 0 ? 0 : votesNotDedicated / this.votesPerPixel;

    this.setState({
      showLogButton: false,
      blockchainLoading: false,
      newVotesPerProject: votesPerProject,
      newVotesNotDedicated: votesNotDedicated,
      voteBarAppliedWidth,
      voteBarFreedUpWidth,
      voteShiftLineLeft: 2 * VOTE_BAR_INNER_MARGIN + voteBarAppliedWidth
    })
  }

  watchVoteChange () {
    this.props.votingInstance.VoteChange().watch((error, event) => {
      this.fetchVoteData();
    })
  }

  render() {
    const { votesPerProject, votesNotDedicated } = this.props;
    if (
        (votesPerProject || votesPerProject === 0) &&
        (votesNotDedicated || votesNotDedicated === 0)
      ) {
      return(
        <React.Fragment>

          <div
            className={this.state.blockchainLoading ? "vote-bar-loading" : "vote-bar"}
            style={{
              width: VOTE_BAR_WIDTH,
              height: VOTE_BAR_HEIGHT,
              borderRadius: VOTE_BAR_RADIUS
            }}>
            <div className="vote-bar-inner-container"
              style={{padding: VOTE_BAR_INNER_MARGIN}}
              ref={node => this.voteBarContainer = node}
              >
              <div className="vote-bar-applied" style={{
                  height: INNER_BAR_HEIGHT,
                  width: this.state.voteBarAppliedWidth,
                  borderRadius: VOTE_BAR_RADIUS
                }}
                ref={node => this.voteBarApplied = node}></div>

              <div className="vote-bar-shift-line" style={{
                  height: VOTE_SHIFT_LINE_HEIGHT,
                  width: VOTE_SHIFT_LINE_WIDTH,
                  left: this.state.voteShiftLineLeft
                }}
                ref={node => this.shiftLine = node}
                onMouseDown={this.handleDragStart}>
                <span className="vote-bar-applied-votes-number">
                  {`${this.state.newVotesPerProject} votes applied`}
                </span>
                <span className="vote-bar-freedup-votes-number">
                  {`${this.state.newVotesNotDedicated} votes freed up`}
                </span>
                <span onClick={this.handleVoteClick(1)} className="vote-bar-add-vote-button">{">"}</span>
                <span onClick={this.handleVoteClick(-1)} className="vote-bar-minus-vote-button">{"<"}</span>
              </div>

              <div className="vote-bar-freedup" style={{
                  height: INNER_BAR_HEIGHT,
                  width: this.state.voteBarFreedUpWidth,
                  borderRadius: VOTE_BAR_RADIUS,
                }}
                ref={node => this.voteBarFreedup = node}></div>
            </div>
          </div>
          {
            this.state.showLogButton &&
            <button className="vote-shift-tool-log-button" onClick={this.handleLogClick}>log</button>
          }
        </React.Fragment>
      );
    } else {
      return []; //make loader
    }
  }
}

VoteShiftTool.defaultProps = {
  votesMockup: {
    votesPerProject: 200,
    votesNotDedicated: 500
  }
};


//CONTAINER
import { connect } from 'react-redux';
import {
  fetchFreeVotes,
  fetchProjectVotes,
  fetchDemoInvestorFreeVotes,
  fetchDemoInvestorProjectVotes,
  voteForProject,
  voteAgainstProject,
  voteAndUpdateProjects,
  demoInvestorVoteAndUpdateProjects
} from '../../../../../actions/chain_actions/votes_actions';
import { updateTransactionModal } from '../../../../../actions/ui_actions';

const mapStateToProps = (state, ownProps) => {
  return {
    projectContract: state.network.projectContract,
    votingToken: state.network.votingTokenInstance,
    votingInstance: state.network.votingInstance,
    account: state.network.account,
    votesNotDedicated: Number(state.chain_data.projectGraph.votes.freeVotes) || 0,
    votesPerProject: Number(state.chain_data.projectGraph.votes[ownProps.selectedProject]) || 0,
    projects: state.chain_data.projects,
    activation: state.network.activationInstance,
    projectLeaderTracker: state.network.projectLeaderTrackerInstance,
    web3: state.network.web3,
    crowdsaleInstance: state.network.crowdsaleInstance,
    inactiveTokenInstance: state.network.inactiveTokenInstance
  }
}

const mapDispatchToProps = dispatch => {
  return {
    fetchFreeVotes: (account, votingToken) => dispatch(fetchFreeVotes(account, votingToken)),
    fetchProjectVotes: (account, projectContract, projectAddress) => dispatch(fetchProjectVotes(account, projectContract, projectAddress)),
    fetchDemoInvestorFreeVotes: () => dispatch(fetchDemoInvestorFreeVotes()),
    fetchDemoInvestorProjectVotes: (projectAddress) => dispatch(fetchDemoInvestorProjectVotes(projectAddress)),
    voteAndUpdateProjects: (account, votes, type, votingInstance, projectAddress, projects, projectLeaderTracker, activation, web3) => {
      return voteAndUpdateProjects(account,
      votes,
      type,
      votingInstance,
      projectAddress,
      projects,
      projectLeaderTracker,
      activation,
      web3)
    },
    updateTransactionModal: modalInfo => dispatch(updateTransactionModal(modalInfo)),
    demoInvestorVoteAndUpdateProjects: (votes, type, selectedProject) => demoInvestorVoteAndUpdateProjects(votes, type, selectedProject)
    // voteForProject: (account, votes, votingInstance, projectAddress) => voteForProject(account, votes, votingInstance, projectAddress),
    // voteAgainstProject: (account, votes, votingInstance, projectAddress) => voteAgainstProject(account, votes, votingInstance, projectAddress)
  }
}


export default connect(mapStateToProps, mapDispatchToProps)(VoteShiftTool);
