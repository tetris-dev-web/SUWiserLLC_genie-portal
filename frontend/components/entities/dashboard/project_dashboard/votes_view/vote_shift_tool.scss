@import "./frontend/util/_variables.scss";

.vote-shift-tool-container {
  position: absolute;
  left: 20px;

  .vote-bar,
  .vote-bar-loading {
    box-sizing: content-box;
    border: 1px solid $tan;
    background-color: $tan;
    margin: 40px 20px 80px 20px;
    position: relative;
    z-index: 20;

    .vote-bar-inner-container {
      display: flex;
      justify-content: space-between;

      .vote-bar-applied {
        background-color: $teal;
      }

      .vote-bar-shift-line {
        cursor: ew-resize;
        background-color: $rosy-brown;
        position: absolute;
        top: -10px;

        .vote-bar-applied-votes-number {
          position: absolute;
          left: -45px;
          top: 40px;
          width: 50px;
        }

        .vote-bar-freedup-votes-number {
          position: absolute;
          right: -55px;
          top: 40px;
          width: 50px;
        }

        .vote-bar-add-vote-button {
          cursor: pointer;
          position: absolute;
          top: -20px;
          left: 11px;
          font-size: 20px;
        }

        .vote-bar-minus-vote-button {
          cursor: pointer;
          position: absolute;
          top: -20px;
          right: 12px;
          font-size: 20px;
        }
      }

      .vote-bar-freedup {
        background-color: #fff;
      }
    }
  }

  .vote-bar-loading {
    background-color: #fff;

    .vote-bar-inner-container {
      .vote-bar-applied {
        background-color: #fff;
        border: 1px solid $teal;
      }

      .vote-bar-freedup {
        border: 1px solid $rosy-brown;
      }
    }
  }

  .vote-shift-tool-log-button {
    position: absolute;
    right: -5px;
    height: 27px;
    top: 40px;
    z-index: 10;
    color: $tan;
    background-color: #fff;
    border: 1px solid $light-gray;
    border-radius: 0 8px 8px 0;

    &:hover {
      background-color: $darkblue;
      color: #fff;
      border-color: $darkblue;
    }
  }
}
