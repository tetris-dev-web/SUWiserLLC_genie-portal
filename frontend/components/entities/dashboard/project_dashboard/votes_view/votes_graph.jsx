import React from 'react';
import * as d3 from 'd3';
import VotesViewCapitalRaised from './votes_view_capital_raised/votes_view_capital_raised';
import VotesViewPitchedProjects from './votes_view_pitched_projects/votes_view_pitched_projects';
import VoteShiftTool from './vote_shift_tool';
import Loader from '../../loader/loader';
import './votes_graph.scss';

class VotesGraph extends React.Component {
  constructor() {
    super();

    this.state = {
      selectedProject: null,
      componentVisible: "invisible",
      voteViewOpen: false,
    };

    this.margin = { top: 20, right: 50, bottom: 30, left: 50 };
    this.SVGWidth = (960 - this.margin.left - this.margin.right) * .5;
    this.SVGHeight = 600 * .5;
    this.timeWidth = (960 - this.margin.left - this.margin.right) * .5;
    this.watchTokenPurchase = this.watchTokenPurchase.bind(this);
    this.watchProjectPitch = this.watchProjectPitch.bind(this);
  }

  componentDidMount() {
    console.log("wait",this.props.wait);
    setTimeout(() => {
      this.setState({componentVisible: ""});
    }, this.props.wait);


    if (!this.props.projectsLoaded) {
      this.props.fetchSharedProjectGraphData();
    }

    this.props.fetchCapitalHistory(this.props.crowdsaleInstance);
    this.watchProjectPitch();
    this.watchTokenPurchase();
  }

  // componentDidUpdate(prevProps) {
  //   const prevLineData = prevProps.lineData;
  //   const { lineData, updateTimeAxis, startTime, endTime } = this.props;
  //   if (prevLineData !== lineData) {
  //     updateTimeAxis(startTime, endTime)
  //   }
  // }

  watchProjectPitch () { //event listener for pitched projects // get project from database and integrate into store
    const { projectFactoryInstance, projectContract } = this.props;
    projectFactoryInstance.ProjectPitch().watch((error, event) => {
      const address = event.args.projectAddress;
      // const id = event.args.projectId;
      this.props.fetchProject(address);
    });
  }

  watchTokenPurchase () {
    console.log("watching purchase")
    this.props.crowdsaleInstance.TokenPurchase().watch((error, event) => {
      this.props.receiveTokenPurchase({blockNumber: event.blockNumber, value: Number(event.args.value)});
      this.props.notifyTransactionCompletion({notification: "Your token purchase transaction has been mined to the blockchain."});
    })
  }

  createScales(){
    const { capitalBeingRaised, capitalDeployed, capitalTotal, pitchedProjectsValuationMinMax, allProjectsValuationMinMax } = this.props;
    const { startTime, endTime } = this.props;
    return {
      SVGHeightScale: d3.scaleLinear()
        .range([0, this.SVGHeight])
        .domain([0, capitalDeployed + Math.max(pitchedProjectsValuationMinMax[1], capitalBeingRaised)]),
      SVGYScale: d3.scaleLinear()
        .range([this.SVGHeight, 0])
        .domain([0, capitalDeployed + Math.max(pitchedProjectsValuationMinMax[1], capitalBeingRaised)]),
      SVGTimeXScale: d3.scaleLinear()
        .domain([startTime, endTime])
        .range([0, this.timeWidth]),
      circleScale: d3.scaleLinear()
        .domain(allProjectsValuationMinMax)
        .range([5, 10])
    };
  }

  dataHasLoaded(){
    return this.props.lineData;
  }
  renderGraph() {
    const { SVGHeightScale, SVGYScale, SVGTimeXScale, circleScale } = this.createScales();
    const { selectedProject, componentVisible } = this.state;
    const { startTime, endTime } = this.props
    let button
    if(!this.state.voteViewOpen){
      button = (<button
                className="voteBreakDownButtonOpen bounceOnHover"
                onClick={() => this.setState({voteViewOpen: !this.state.voteViewOpen})}>
                projects under consideration
                </button>)
    }else if(selectedProject){
      button = ""
    }else{
      button = (<button
                className="voteBreakDownButtonClose bounceOnHover"
                onClick={() => this.setState({voteViewOpen: !this.state.voteViewOpen})}>
                X
                </button>)
    }

    return (
      <div className={`votes-graph ${componentVisible}`}>
        <div className="vote-shift-tool-container"
          ref={node => this.voteShiftTool = node}>
          {
            selectedProject &&
            <VoteShiftTool selectedProject={selectedProject.address}/>
          }
        </div>
        {
          button
        }

        <svg className="votes-view-svg"
          preserveAspectRatio="xMinYMin meet"
          viewBox="0 0 960 300">
          <VotesViewCapitalRaised
            {...this.props}
            {...this.state}
            margin={this.margin}
            SVGYScale={SVGYScale}
            SVGHeightScale={SVGHeightScale}
            SVGTimeXScale={SVGTimeXScale}
            circleScale={circleScale}
            startTime={startTime}
            endTime={endTime}
          />

          {
          !this.state.voteViewOpen?
          ""
          : <VotesViewPitchedProjects
              {...this.props}
              {...this.state}
              margin={this.margin}
              SVGYScale={SVGYScale}
              SVGHeightScale={SVGHeightScale}
              SVGWidth={this.SVGWidth}
              circleScale={circleScale}
              voteShiftTool={this.voteShiftTool}
              toggleSelectedProject={selectedProject => this.setState({selectedProject})}
            />
           }
        </svg>
      </div>
    );
  }

  render() {
    return this.dataHasLoaded() ? this.renderGraph() : <Loader/>;
  }
}

export default VotesGraph;
