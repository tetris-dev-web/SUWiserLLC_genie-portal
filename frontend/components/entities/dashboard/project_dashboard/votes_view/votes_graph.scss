@import "./frontend/util/_variables.scss";

.votes-graph {
  position: relative;
  transition: opacity 0.3s;
  visibility: visible;
  opacity: 1;
  text-align: center;
  .voteBreakDownButtonOpen {
    background: $tan;
    border-radius: 15px;
    color: white;
    margin: 25px;
  }
  .voteBreakDownButtonClose {
    background: $tan;
    border-radius: 15px;
    color: white;
    opacity: 0.3;
    margin: 25px;
  }

  &.invisible {
    visibility: hidden;
    opacity: 0;
  }

  .votes-view {
    top: 0;
    position: absolute;
    width: 100%;
  }

  .votes-view-svg {
    width: 100%;
  }

  .ProjectCrossLine {
    stroke: rgb(189, 196, 201);
    stroke-width: 5px;
  }

  .CapitalRaisedPath {
    fill: none;
    stroke: $teal;
    stroke-width: 4px;
  }

  .questionIcon {
    fill: $teal;
  }

  text {
    text-anchor: middle;
    pointer-events: none;
  }

  // .capital-raised-line {
  //   fill: blue;
  //   width: 100px;
  // }

  .votes-view-vote-breakdown-text {
    fill: #fff;
    font-size: 12px;
  }

  .votes-view-circle-text-hovers {
    fill: rgb(189, 196, 201);
  }

  .votes-view-capital-raised-text {
    fill: #fff;
  }
  .votes-view-capital-raised-text-right {
    fill: #fff;
  }

  .votes-view-percentage-breakdown {
    fill: #fff;
    font-size: 12px;
  }

  .votes-view-onhover-text-group {
    text {
      font-size: 12px;
    }
    .votes-view-project-title {
      font-size: 12px;
      fill: white;
    }
  }

  .votes-view-onhover-project-outlines-group {
    rect {
      fill: #fff;
      pointer-events: none;
    }
  }

  .votes-view-project-circle {
    transition: r 0.5s;
    cursor: pointer;

    &:hover {
      r: 12px;
    }
  }
}
