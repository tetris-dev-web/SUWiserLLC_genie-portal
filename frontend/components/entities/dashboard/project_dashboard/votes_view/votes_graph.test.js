import React from 'react'
import {render, unmountComponentAtNode} from 'react-dom'
import {VotesGraph as Graph} from 'Entities/dashboard/project_dashboard/votes_view/votes_graph'
import { shallow, mount, render as eRender } from 'enzyme'


// import {setUp} from './project_graph_d3logic';

function componentSetup(){
  const props = {
    fetchProjects: jest.fn(()=>{
      return Promise.resolve();
    }),
    prop_sample:  {1:0, 2:-30, 3: -10,  4: 0,   5: 0},
  }
  const eWrapper = shallow(<Graph {...props}/>)
  return {
    props,
    eWrapper
  }
}


describe('Project Module Component', () => {
 // it('the data being rendered should match the number of projects in the database', () => {
 //   const { eWrapper} = setup()
 //   // expect(eWrapper.find('section').hasClass('graph-container')).toBe(true)
 //    expect(eWrapper).toMatchSnapshot();
 // })


   it('should match snapshot', () => {
     const { eWrapper} = componentSetup()
     // expect(eWrapper.find('section').hasClass('graph-container')).toBe(true)
      expect(eWrapper).toMatchSnapshot()
   })




})
