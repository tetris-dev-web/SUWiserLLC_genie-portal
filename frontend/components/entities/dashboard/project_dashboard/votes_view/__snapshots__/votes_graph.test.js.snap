// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Module Component should match snapshot 1`] = `
<div
  className="votes-graph"
  style={
    Object {
      "marginTop": NaN,
    }
  }
>
  <svg
    className="votes-view-svg"
    height={NaN}
  >
    <VotesViewCapitalRaised
      fetchProjects={[MockFunction]}
      prop_sample={
        Object {
          "1": 0,
          "2": -30,
          "3": -10,
          "4": 0,
          "5": 0,
        }
      }
      selectedProject={null}
    />
    <VotesViewPitchedProjects
      fetchProjects={[MockFunction]}
      prop_sample={
        Object {
          "1": 0,
          "2": -30,
          "3": -10,
          "4": 0,
          "5": 0,
        }
      }
      selectedProject={null}
      toggleSelectedProject={[Function]}
    />
  </svg>
</div>
`;
